src/main/java/com/azure/cosmos/kafka/connect/sink/BulkWriter.java

```diff
@@ -0,0 +1,146 @@
+        super(maxRetryCount);
+
+        checkNotNull(container, "Argument 'container' can not be null");
+        this.cosmosContainer = container;
+        this.partitionKeyDefinition = container.read().getProperties().getPartitionKeyDefinition();
+    }
+
+    /***
+     * Bulk write the sink records.
+     *
+     * @param sinkRecords the sink records needs to be written.
+     * @return the list of sink write failed operations.
+     */
+    protected SinkWriteResponse writeCore(List<SinkRecord> sinkRecords) {
+
+        SinkWriteResponse sinkWriteResponse = new SinkWriteResponse();
+        if (sinkRecords == null || sinkRecords.isEmpty()) {
+            return sinkWriteResponse;
+        }
+
+
+        List<CosmosItemOperation> itemOperations = new ArrayList<>();
+        for (SinkRecord sinkRecord : sinkRecords) {
+            CosmosItemOperation cosmosItemOperation = CosmosBulkOperations.getUpsertItemOperation(
+                    sinkRecord.value(),
+                    this.getPartitionKeyValue(sinkRecord.value()),
+                    new SinkOperationContext(sinkRecord));
+
+            itemOperations.add(cosmosItemOperation);
+        }
+
+        Iterable<CosmosBulkOperationResponse<Object>> responseList = cosmosContainer.executeBulkOperations(itemOperations);
+
+        // Non-transient exceptions will be put in the front of the list
+        for (CosmosBulkOperationResponse bulkOperationResponse : responseList) {
+            SinkOperationContext context = bulkOperationResponse.getOperation().getContext();
+            SinkRecord sinkRecord = context.getSinkRecord();
+
+            if (bulkOperationResponse.getException() != null
+                    || bulkOperationResponse.getResponse() == null
+                    || !bulkOperationResponse.getResponse().isSuccessStatusCode()) {
+
+                BulkOperationFailedException exception = handleNonSuccessfulStatusCode(
+                        bulkOperationResponse.getResponse(),
+                        bulkOperationResponse.getException(),
+                        bulkOperationResponse.getOperation().getContext());
+
+                if (ExceptionsHelper.canBeTransientFailure(exception)) {
+                    sinkWriteResponse.getFailedRecordResponses().add(new SinkOperationFailedResponse(sinkRecord, exception));
+                } else {
+                    sinkWriteResponse.getFailedRecordResponses().add(0, new SinkOperationFailedResponse(sinkRecord, exception));
+                }
+            } else {
+                sinkWriteResponse.getSucceededRecords().add(sinkRecord);
+            }
+        }
+
+        return sinkWriteResponse;
+    }
+
+    private PartitionKey getPartitionKeyValue(Object recordValue) {
+        checkArgument(recordValue instanceof Map, "Argument 'recordValue' is not valid map format.");
+
+        //TODO: examine the code here for sub-partition
+        String partitionKeyPath = StringUtils.join(this.partitionKeyDefinition.getPaths(), "");
```

Comment:
This is all not future safe for subpartitioning - worth fixing this now? Or as part of subpartitioning separately?